### PR TITLE
nixos/networkmanager: allow iwd as Wi-Fi backend

### PR DESCRIPTION
**Pull request by colemickens**
Replaces #51803. Please see it for earlier discussion. I think all I did was rebase and fix merge conflicts. As requested by @dtzWill.

Works well for me. Connecting and disconnecting via `nmtui` is remarkably faster than with wpa_supplicant (or, at least, without this change and using IWD).

### CODE REVIEW

**Comment by colemickens** [comment]
Should this target staging, maybe?

**Comment by worldofpeace** [comment]
> Should this target staging, maybe?

It shouldn't cause a mass rebuild since the feature is already enabled, changes don't touch the networkmanager expression.

File: nixos/modules/services/networking/networkmanager.nix [hunk 1]
```diff
@@ -35,6 +35,8 @@ let
 
     [device]
     wifi.scan-rand-mac-address=${if cfg.wifi.scanRandMacAddress then "yes" else "no"}
+    ${optionalString (cfg.wifi.backend != null)
+      ''wifi.backend=${cfg.wifi.backend}''}
```
Comment (worldofpeace) [inline nixos/modules/services/networking/networkmanager.nix]:
```suggestion
      "wifi.backend=${cfg.wifi.backend}"}
```

File: nixos/modules/services/networking/networkmanager.nix [hunk 2]
```diff
@@ -177,7 +179,7 @@ in {
       basePackages = mkOption {
         type = types.attrsOf types.package;
         default = { inherit (pkgs)
-                            networkmanager modemmanager wpa_supplicant crda
```
Comment (worldofpeace) [inline nixos/modules/services/networking/networkmanager.nix]: Every reference to `networking.networkmanager.basePackages` should drop `wpa_supplicant`.
(gnome3, etc.)

File: nixos/modules/services/networking/networkmanager.nix [hunk 3]
```diff
@@ -386,6 +394,12 @@ in {
           networking.networkmanager.dns = "dnsmasq"
         '';
       }
+      { assertion = !config.networking.wireless.iwd.enable;
+        message = ''
+          Only one option to enable iwd is allowed: networking.wireless.iwd.enable
```
Comment (worldofpeace) [inline nixos/modules/services/networking/networkmanager.nix]: Can you put  markdown code blocks around the options?

File: nixos/modules/services/networking/networkmanager.nix [hunk 4]
```diff
@@ -232,6 +234,12 @@ in {
       wifi = {
         macAddress = macAddressOpt;
 
+        backend = mkOption {
+          type = types.enum ["wpa_supplicant" "iwd"];
+          default = "wpa_supplicant";
+          description = "WiFi Backend to be used by NetworkManager";
```
Comment (worldofpeace) [inline nixos/modules/services/networking/networkmanager.nix]:
```suggestion
          description = "WiFi Backend to be used by NetworkManager.";
```

File: nixos/modules/services/networking/networkmanager.nix (hunk 2 above)
Comment (colemickens) [reply nixos/modules/services/networking/networkmanager.nix]: Hm, why were so many places re-setting the basePackages? Just trying to understand why/what the change is...

Comment (worldofpeace) [reply nixos/modules/services/networking/networkmanager.nix]: It appears to be a hack so networkmanager can use the gnome version of the packages from the `gnome3` attrset. GNOME type libs are pretty popular so, they're built with those features.

Comment (worldofpeace) [reply nixos/modules/services/networking/networkmanager.nix]: services.xserver.desktopManager.xfce4-14 as well

**Comment by worldofpeace** [comment]
So i tried using this and it appears there is some config that should be shared between the iwd module.
This probably worked for because maybe you were using it before.

```diff
diff --git a/nixos/modules/services/networking/networkmanager.nix b/nixos/modules/services/networking/networkmanager.nix
index 0dc4aec95ab..1d4c1dd585e 100644
--- a/nixos/modules/services/networking/networkmanager.nix
+++ b/nixos/modules/services/networking/networkmanager.nix
@@ -526,5 +526,11 @@ in {
       optional cfg.enableStrongSwan pkgs.strongswanNM ++ cfg.packages;
 
     services.udev.packages = cfg.packages;
+
+    systemd.services.iwd.wantedBy = mkIf (cfg.wifi.backend == "iwd") [ "multi-user.target" ];
+
+    systemd.tmpfiles.rules = mkIf (cfg.wifi.backend == "iwd") [
+      "d /var/lib/iwd 0700 root root -"
+    ];
   };
 }

```

with that it worked on my system.

I have to ask if they're really mutually exclusive? How does NetworkManager use iwd?

**Comment by worldofpeace** [comment]
You should also include #68213 for networkmanager.

**Comment by colemickens** [comment]
Hm, I thought the assert that's in this change was specifically because NM controls the IWD lifecycle when IWD is used with NM. I may have had the `/var/lib/iwd` directory because I had briefly directly enabled `iwd`, but my setup is working fine without the extra `wantedBy` you have suggested, again (I thought), due to NM controlling the IWD lifecycle in this scenario.

Do you want me to pull the commit from #68213 in here, or just review it and make this PR depend on merging it first?

**Comment by worldofpeace** [comment]
> Hm, I thought the assert that's in this change was specifically because NM controls the IWD lifecycle when IWD is used with NM. I may have had the /var/lib/iwd directory because I had briefly directly enabled iwd, but my setup is working fine without the extra wantedBy you have suggested, again (I thought), due to NM controlling the IWD lifecycle in this scenario.

Yeah, I think because of that the state directories would already have been ready.
The wantedBy is already in the iwd module so that's why it's there.

> Do you want me to pull the commit from #68213 in here, or just review it and make this PR depend on merging it first?

I'm not sure if we're going the route of enabling the `iwd` module conditionally within the networkmanager module (I think it could be possible) or duplicating it.
Looking at the code here, using `cfg.packages` actually adds the packages to the same options as the iwd module does already. So I think it would be fine to just enable iwd within the networkmanager module.

**Comment by worldofpeace** [comment]
We discussed in depth on irc
* https://logs.nix.samueldr.com/nixos/2019-09-06#2563827;

**Comment by worldofpeace** [comment]
@colemickens, new change got merged that added another variable to check into the mix.
I added https://github.com/NixOS/nixpkgs/pull/68147/commits/b07d6f5d2b2354ab753cdfec67396ecad9aa36f4 for this, I hope you don't mind.

**Comment by colemickens** [comment]
I had to do some weird merging fixups when I cherrypicked the ell/iwd-bump and then these two commits, but it seems to work fine. Should I squash and repush these, @worldofpeace ?

**Comment by worldofpeace** [comment]
> I had to do some weird merging fixups when I cherrypicked the ell/iwd-bump and then these two commits, but it seems to work fine. Should I squash and repush these, @worldofpeace ?

Sure, I'll also request the review of one more person :+1:

**Comment by abbradar** [comment]
What does `wifi.backend = null` do? Seems it means "don't manage wireless by NetworkManager" but then `delegateWireless` also checks that `unmanaged` isn't empty. What would happen with `cfg.wifi.backend = null; cfg.unmanaged = []; networking.wireless.enable = true;`?

**Comment by worldofpeace** [comment]
Oh, we should update the assertion.

**Comment by worldofpeace** [comment]
I don't think having `wifi.backend` unset makes it not manage wpa_supplicant, `unmanaged = [ "*"];` will probably do this. So I can drop that being `null`.

Though I think https://github.com/NixOS/nixpkgs/commit/b4044a3f2a56b95e76e8091a8e5d9c52fa24a3b6 doesn't work as expected.
With
```
networking.networkmanager.enable = true;  
networking.wireless.enable = true;
```

I don't get the assertion message.

**Comment by colemickens** [comment]
@worldofpeace: I'm a bit confused with the change you were trying to make that apparently isn't quite right. I'm probably going to hold off until the other referenced PR hits nixos-unstable, and then I can take a look again. Do you want to have a crack at fixing the assert message? I only have one wlan interface and am short of time to test and push this forward for potentially a couple of weeks. Thanks.

**Comment by worldofpeace** [comment]
It would have been no-op until https://github.com/NixOS/nixpkgs/commit/b4044a3f2a56b95e76e8091a8e5d9c52fa24a3b6 landed. We could revert it on master to fix the issue, as the assertion for that was broken. And then we can revert my fixup commit and merge.
Though I'm short on time with other things currently.

**Comment by JohnAZoidberg** [comment]
I'll check it out until tomorrow how to reconcile this with b4044a3.

**Comment by JohnAZoidberg** [comment]
I'm not sure why

```
networking.networkmanager.enable = true;  
networking.wireless.enable = true;
```

works for you, @worldofpeace. :thinking: 
The assertion clearly rules that out. Do you have `networking.networkmanager.unmanaged` set to a non-empty list somewhere?

With the new changes here, though, I'd add this diff to the PR:

```diff
--- i/nixos/modules/services/networking/networkmanager.nix
+++ w/nixos/modules/services/networking/networkmanager.nix
@@ -394,10 +394,12 @@ in {
   config = mkIf cfg.enable {

     assertions = [
-      { assertion = config.networking.wireless.enable == true -> cfg.unmanaged != [];
+      { assertion = config.networking.wireless.enable ->
+          (cfg.unmanaged != [] || cfg.wifi.backend == null);
         message = ''
           You can not use networking.networkmanager with networking.wireless.
-          Except if you mark some interfaces as <literal>unmanaged</literal> by NetworkManager.
+          Except if you mark some interfaces as `unmanaged` by NetworkManager
+          or set the wifi backend to `null`.
         '';
       }
       { assertion = !dynamicHostsEnabled || (dynamicHostsEnabled && cfg.dns == "dnsmasq");
```

**Comment by worldofpeace** [comment]
Huh, I tested again on a running system and I get the assertion message. I guess you can't test this with a vm.

Also, I'm not sure if networkmanager has a concept of not using any backend. So allowing null as a value doesn't make sense. What actually matters is what's set as unmanaged.

**Comment by worldofpeace** [comment]
@JohnAZoidberg I believe this should be correct now.

**Review by JohnAZoidberg — APPROVED** [review]
Yes, you're right. NM seems to use supplicant if no `wifi.backend` is given.

**Comment by worldofpeace** [comment]
Interesting, 

> This works for me. But `dhclient` gives me the following error:
> 
> ```
> Can't bind to dhcp address: Address already in use
> Please make sure there is no other dhcp server
> running and that there's no entry for dhcp or
> bootp in /etc/inetd.conf.   Also make sure you
> are not running HP JetAdmin software, which
> includes a bootp server.
> 
> If you think you have received this message due to a bug rather
> than a configuration issue please read the section on submitting
> bugs on either our web page at www.isc.org or in the README file
> before submitting a bug.  These pages explain the proper
> process and the information we find helpful for debugging.
> 
> exiting.
> ```
> 
> So I guess `dhclient` shouldn't be running when using `iwd`?

We currently default to `internal`. Also, I'm not sure if the problem would be them conflicting in any way. iwd has nothing to do with dhcp.

**Comment by xaverdh** [comment]
> We currently default to `internal`. Also, I'm not sure if the problem would be them conflicting in any way. iwd has nothing to do with dhcp.

Actually, I think `dhclient` is the default if I read the code on [master](https://github.com/NixOS/nixpkgs/blob/6d2e24e80e8a0cc5318bc93d5c6b565008aea245/nixos/modules/services/networking/networkmanager.nix#L202) correctly.
Also from the [Arch Wiki](https://wiki.archlinux.org/index.php/Iwd):
```Since version 0.19, iwd can assign IP address(es) and set up routes using a built-in DHCP client or with static configuration.```
Although according to https://git.kernel.org/pub/scm/network/wireless/iwd.git/commit/?id=199e647ffde38584d5bab609de38bea973082ffe iwd's internal dhcp should be off by default.

**Comment by worldofpeace** [comment]
@xaverdh Oh, that change hasn't hit master yet https://github.com/NixOS/nixpkgs/commit/1ff3a0c4986befbda79b108103496fa4f846fd9a.

And it does seem it has this function. I don't think the module for iwd handles any config file so I bet our issue is right there.

**Comment by xaverdh** [comment]
Indeed with internal `dhcp` everything works just fine.

**Comment by jluttine** [comment]
For your information, this pull request fixes https://github.com/NixOS/nixpkgs/issues/69368 (thanks @xaverdh). (Or not sure if fixing the issue is correctly said, but at least offers one workaround.)

**Comment by asymmetric** [comment]
OT: what's the advantage of using iwd as a NM backend as opposed to just using `iwctl`?

**Comment by jluttine** [comment]
Not familiar with these, so I might speak nonsense, but AFAIU, NetworkManager has everything in one place. So you can manage wired&wireless connections and VPN from one app/applet/trayicon.

**Comment by worldofpeace** [comment]
Rebased as I've removed the `basePackages` https://github.com/NixOS/nixpkgs/pull/70041 code which was unneeded.

**Pull request merged**